### ISSUE TITLE
Templated GEMM references C when beta == 0.0.

### ISSUE THREAD

**Issue by rileyjmurray** [icl-utk-edu/blaspp#39]
The fully templated GEMM code (below) references the input value of ``C(i,j)`` even when ``beta == 0.0``. As a result, if ``C`` isn't initialized and contains a NaN, then that NaN can still be propagated through to the output when ``beta == 0.0``.

https://github.com/icl-utk-edu/blaspp/blob/e954a9bfa4187835308e507735887d91f1132d83/include/blas/gemm.hh#L166-L265

@weslleyspereira, recently I suggested that tests for this situation be added to your BLAS stress tests. Once you add such tests, it would be good to run against BLAS++. It might be that other fully-templated functions in BLAS++ make similar mistakes.

**Comment by weslleyspereira** [comment]
Hi. Good catch, @rileyjmurray !

This test is already in testBLAS. See
https://github.com/tlapack/testBLAS/blob/b177c5b3ae3dbbaece254727e036f487d2271506/src/test_corner_cases.cpp#L993-L997
You can use testBLAS to test the templates from BLAS++. It is prepared for that.

That said, I am not sure what fix should be considered in BLAS++. In [\<T\>LAPACK](https://github.com/tlapack/tlapack) we adopted a different strategy. There, we have 2 interfaces for gemm:
1. `C := alpha*A*B+beta*C`
2. `C := alpha*A*B`

See https://github.com/tlapack/tlapack/blob/master/include/tlapack/blas/gemm.hpp

**Comment by rileyjmurray** [comment]
The fix in BLAS++ should just be to write ``C(i, j) = 0.0`` for all ``(i, j)`` as a preliminary step if ``beta`` is zero. This is the only thing that would conform with the BLAS standard, as I understand it. It could even be performed at the start of the function and subsequent code could be left unchanged relative to its current form (note: that would mean $O(n^2)$ multiplications are wasted when ``beta == 0.0``, but that's _really_ not something we should care about).

**Comment by weslleyspereira** [comment]
I agree with you. Moreover, this fix should be applied to all lv3 BLAS routines.

**Comment by mgates3** [comment]
Depending on the case, reference BLAS either [sets `C = 0` before](https://github.com/Reference-LAPACK/lapack/blob/master/BLAS/SRC/dgemm.f#L299-L302), or sets [`C = alpha*A*B` after](https://github.com/Reference-LAPACK/lapack/blob/master/BLAS/SRC/dgemm.f#L325-L329) (without `beta*C`). Not really sure why they chose 2 different implementations. The later one has less memory traffic (one less write of C and read of C).

Also, if `alpha = 0`, it should just do `beta*C` as [reference BLAS does](https://github.com/Reference-LAPACK/lapack/blob/master/BLAS/SRC/dgemm.f#L271-L289), no need to do O(n^3) work in `alpha*A*B`. Though I argue that using `gemm` to scale or initialize a matrix is an abuse of `gemm`; use `lascl` or `laset`.

**Comment by weslleyspereira** [comment]
> Also, if `alpha = 0`, it should just do `beta*C` as [reference BLAS does](https://github.com/Reference-LAPACK/lapack/blob/master/BLAS/SRC/dgemm.f#L271-L289), no need to do O(n^3) work in `alpha*A*B`. Though I argue that using `gemm` to scale or initialize a matrix is an abuse of `gemm`; use `lascl` or `laset`.

I agree with your argument. One should use `lascl` or `laset` if alpha is known to be zero. The current implementation of gemm in \<T\>LAPACK will do the O(n^3) operation `alpha*A*B` even if alpha is zero.

**Comment by weslleyspereira** [comment]
I noticed that I wasn't testing BLAS++ with the option ``. Now the automatic tests in testBLAS are able to catch the issues related to this thread. See https://github.com/tlapack/testBLAS/actions/runs/4484673095/jobs/7885422971

**Comment by mgates3** [comment]
@weslleyspereira Which option? Your post says: option ``.

**Comment by weslleyspereira** [comment]
> @weslleyspereira Which option? Your post says: option ``.

I'm sorry. I meant `BLAS_USE_TEMPLATE`.

**Comment by mgates3** [comment]
Fixed in #99.
